### PR TITLE
Add `dnf_module` resource

### PR DESCRIPTION
**Pull request by detjensrobert**
# Description

Chef does not have a way to manage DNF modules on C8+, so this does that.

## Issues Resolved
 
(none)

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

### CODE REVIEW

File: resources/dnf_module.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+          description: 'Name of the module to install'
+
+property :options, [String, Array],
+          coerce: proc { |x| Array(x) },
+          default: [],
+          description: 'Any additional options to pass to DNF'
+
+action_class do
+  def enabled_modules
+    # look for modules with [e]
```
Comment (Xorima) [inline resources/dnf_module.rb:17]: what is significant about [e]? Could be useful to have that in the comment

File: resources/dnf_module.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    raw_output = shell_out!('dnf -q module list --enabled').stdout.split("\n")
+    raw_output.keep_if { |l| l.match? /\[e\] / }
+    raw_output.map { |l| "#{l.split[0]}:#{l.split[1]}" }
+  end
+
+  def installed_modules
+    # look for modules with [i]
```
Comment (Xorima) [inline resources/dnf_module.rb:24]: what is significant about [i]? Could be useful to have that in the comment

File: test/fixtures/cookbooks/yum_test/metadata.rb [hunk 3]
```diff
@@ -1,8 +1,3 @@
 name 'yum_test'
```
Comment (Xorima) [inline test/fixtures/cookbooks/yum_test/metadata.rb:1]: If we are in here anyway can we rename this folder and this cookbook to be `test` in line with our normal conventions please?

File: resources/dnf_module.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+  def list_modules(type)
+    raw_output = shell_out!('dnf -q module list').stdout.split("\n")
+    raw_output.keep_if { |l| l.match? /\[#{type}\]/ }
+    raw_output.map { |l| "#{l.split[0]}:#{l.split[1]}" }[0..-2] # remove Hint: line from end
+  end
+
+  def enabled_modules
+    # extract modules from the rest of the output -- the lines with [e] (for enabled)
+    list_modules 'e'
```
Comment (bmhughes) [inline resources/dnf_module.rb:28]: Lets stick some parentheses on these method calls

File: resources/dnf_module.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  end
+
+  def disabled_modules
+    # extract modules from the rest of the output -- the lines with [x] (for disabled)
+    # disable disables all versions of the stream, so add entry without :version
+    dl = list_modules 'x'
```
Comment (bmhughes) [inline resources/dnf_module.rb:34]: Parentheses

File: resources/dnf_module.rb [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    dl + dl.map { |m| m.split(':').first }.uniq
+  end
+
+  def installed_modules
+    # extract modules from the rest of the output -- the lines with [i] (for installed)
+    list_modules 'i'
```
Comment (bmhughes) [inline resources/dnf_module.rb:40]: Parentheses

File: resources/dnf_module.rb [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+  def opts
+    new_resource.options.join(' ')
+  end
+end
+
+action :switch_to do
+  return unless supported?
+
+  unless enabled_modules.include? new_resource.module_name
```
Comment (bmhughes) [inline resources/dnf_module.rb:51]: Parentheses

File: resources/dnf_module.rb [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    converge_by "switch to #{new_resource.module_name}" do
+      shell_out!("dnf -qy module switch-to #{opts} '#{new_resource.module_name}'")
+    end
+  end
+end
+
+action :enable do
+  return unless supported?
+
+  unless enabled_modules.include? new_resource.module_name
```
Comment (bmhughes) [inline resources/dnf_module.rb:61]: Parentheses

File: resources/dnf_module.rb [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    converge_by "enable #{new_resource.module_name}" do
+      shell_out!("dnf -qy module enable #{opts} '#{new_resource.module_name}'")
+    end
+  end
+end
+
+action :disable do
+  return unless supported?
+
+  unless disabled_modules.include? new_resource.module_name
```
Comment (bmhughes) [inline resources/dnf_module.rb:71]: Parentheses

File: resources/dnf_module.rb [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    converge_by "disable #{new_resource.module_name}" do
+      shell_out!("dnf -qy module disable #{opts} '#{new_resource.module_name}'")
+    end
+  end
+end
+
+action :install do
+  return unless supported?
+
+  unless installed_modules.include? new_resource.module_name
```
Comment (bmhughes) [inline resources/dnf_module.rb:81]: Parentheses

File: resources/dnf_module.rb [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    converge_by "install #{new_resource.module_name}" do
+      shell_out!("dnf -qy module install #{opts} '#{new_resource.module_name}'")
+    end
+  end
+end
+
+action :remove do
+  return unless supported?
+
+  if installed_modules.include? new_resource.module_name
```
Comment (bmhughes) [inline resources/dnf_module.rb:91]: Parentheses

**Review by bmhughes — DISMISSED** [review]

File: documentation/dnf_module.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+- `:install` - Enable a module stream and install its packages
+- `:remove` - Disable a module stream and remove its packages
+- `:reset` - Unset module state and remove packages not in the default streams **(this action is not idempotent)**
+
+## Properties
+
+| Name          | Type                | Default       | Description                           |
+| ------------- | ------------------- | ------------- | ------------------------------------- |
+| `module_name` | `String`            | Resource name | Name of the module to install.        |
+| `options`     | `String` or `Array` |               | Any additional options to pass to DNF |
```
Comment (ramereth) [inline documentation/dnf_module.md:23]: Please add an example.

**Review by ramereth — CHANGES_REQUESTED** [review]

**Review by ramereth — APPROVED** [review]

**Comment by kitchen-porter** [comment]
Released as: [7.2.0](https://github.com/sous-chefs/yum/releases/tag/7.2.0)

File: resources/dnf_module.rb [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+          description: 'Name of the module to install'
+
+property :options, [String, Array],
+          coerce: proc { |x| Array(x) },
+          default: [],
+          description: 'Any additional options to pass to DNF'
+
+action_class do
+  def supported?
+    (platform_family?('rhel') && node['platform_version'] >= 8) || platform?('fedora')
```
Comment (lordkage) [inline resources/dnf_module.rb:17]: Shouldn't `node['platform_version'] >= 8` be `node['platform_version'].to_i >= 8` ?

Comment (tas50) [reply resources/dnf_module.rb:17]: On 15.3+ you are going to need to .to_i it. In the latest versions you don't need to anymore.

Comment (detjensrobert) [reply resources/dnf_module.rb:17]: Whoops, how did I miss that in my original pr.
Fix in #197.

Comment (lordkage) [reply resources/dnf_module.rb:17]: Thank you

**Pull request merged**